### ISSUE TITLE
$1 gets clobbered in Net::hostent::gethost

### ISSUE THREAD

**Issue by eserte** [Perl/perl5#19017]
`$1` and other capture variables cannot be used in a `gethost` call of `Net::hostent`:
```
$ perl -w -MNet::hostent=gethost -e '"127.0.0.1"=~/(.*)/; warn gethost $1'
Use of uninitialized value in subroutine entry at /opt/perl-5.12.3/lib/5.12.3/Net/hostent.pm line 58.
Use of uninitialized value $addr in gethostbyaddr at /opt/perl-5.12.3/lib/5.12.3/Net/hostent.pm line 52.
Warning: something's wrong at -e line 1.
```
Workaround: assign $1 to another variable and use that one:
```
$ perl -w -MNet::hostent=gethost -e '"127.0.0.1"=~/(.*)/; $x = $1; warn gethost $x'
Net::hostent=ARRAY(0x5640902fda28) at -e line 1.
```
Alternative: remove the "aliasness" (?) by stringifying the value:
```
$ perl -w -MNet::hostent=gethost -e '"127.0.0.1"=~/(.*)/; warn gethost "$1"'
Net::hostent=ARRAY(0x556681ba8a28) at -e line 1.
```
But a better workaround would be if this happened in Net/hostent.pm.

This problem seems to exist forever (tried 5.8.8 and 5.34.0).

Informal patch to fix the problem:
```
diff --git a/lib/Net/hostent.pm b/lib/Net/hostent.pm
index 8831919fb4..77e126ad9a 100644
--- a/lib/Net/hostent.pm
+++ b/lib/Net/hostent.pm
@@ -58,11 +58,12 @@ sub gethostbyaddr ($;$) {
 }

 sub gethost($) {
-    if ($_[0] =~ /^\d+(?:\.\d+(?:\.\d+(?:\.\d+)?)?)?$/) {
+    my $addr = shift;
+    if ($addr =~ /^\d+(?:\.\d+(?:\.\d+(?:\.\d+)?)?)?$/) {
        require Socket;
-       &gethostbyaddr(Socket::inet_aton(shift));
+       &gethostbyaddr(Socket::inet_aton($addr));
     } else {
-       &gethostbyname;
+       &gethostbyname($addr);
     }
 }
```

Adding test cases to the test script (it also seems that the test case `gethost($ip_address)` is currently missing:
```
diff --git a/lib/Net/hostent.t b/lib/Net/hostent.t
index 573136c3b1..f1be69e836 100644
--- a/lib/Net/hostent.t
+++ b/lib/Net/hostent.t
@@ -19,7 +19,7 @@ BEGIN {
     }
 }

-use Test::More tests => 7;
+use Test::More tests => 10;

 BEGIN { use_ok 'Net::hostent' }

@@ -48,6 +48,14 @@ ok(defined $i,  "gethostbyaddr('127.0.0.1')") ||

 is( inet_ntoa($i->addr), "127.0.0.1",   'addr from gethostbyaddr' );

+$i = gethost("127.0.0.1");
+ok(defined $i,  "gethost('127.0.0.1')");
+is( inet_ntoa($i->addr), "127.0.0.1",   'addr from gethost' );
+
+"127.0.0.1" =~ /(.*)/;
+$i = gethost($1);
+ok(defined $i, 'gethost on capture variable');
+
 # need to skip the name comparisons on Win32 because windows will
 # return the name of the machine instead of "localhost" when resolving
 # 127.0.0.1 or even "localhost"
```

**Comment by jkeenan** [comment]
> `$1` and other capture variables cannot be used in a `gethost` call of `Net::hostent`:
> 

I have created a branch from these patches: smoke-me/jkeenan/eserte/gh-19017-net-hostent-20210804

... and a pull request:

https://github.com/Perl/perl5/pull/19023

Please review.

**Comment by jkeenan** [comment]
> > `$1` and other capture variables cannot be used in a `gethost` call of `Net::hostent`:
> 
> I have created a branch from these patches: smoke-me/jkeenan/eserte/[gh-19017](https://github.com/Perl/perl5/issues/19017)-net-hostent-20210804
> 
> ... and a pull request:
> 
> #19023
> 
> Please review.

@eserte, @TonyCoz approved the pull request I created from the "informal" patches you submitted in this ticket.  Is there anything you would like to do to make them "formal"?

Thank you very much.
Jim Keenan

**Comment by jkeenan** [comment]
> > > `$1` and other capture variables cannot be used in a `gethost` call of `Net::hostent`:
> > 

> 
> @eserte, @tonycoz approved the pull request I created from the "informal" patches you submitted in this ticket. Is there anything you would like to do to make them "formal"?
> 

Having received no further feedback, I merged #19023 into blead in a4c6f3fb2a4bc5e20603fc62259532eae0142365.  Closing ticket.
